infoschema v1 and v2 are inconsistent during TestDisaggregatedTiFlashGeneratedColumn

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

This issue is similar to #55442. Run TestDisaggregatedTiFlashGeneratedColumn using #55030.

<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)

Test passed

### 3. What did you see instead (Required)

```
--- FAIL: TestDisaggregatedTiFlashGeneratedColumn (0.51s)
panic: inconsistent infoschema [recovered]
        panic: inconsistent infoschema [recovered]
        panic: inconsistent infoschema

goroutine 23 [running]:
testing.tRunner.func1.2({0x1070259a0, 0x10795bfe0})
        /Users/joechenrh/deps/go/src/testing/testing.go:1631 +0x1c4
testing.tRunner.func1()
        /Users/joechenrh/deps/go/src/testing/testing.go:1634 +0x33c
panic({0x1070259a0?, 0x10795bfe0?})
        /Users/joechenrh/deps/go/src/runtime/panic.go:770 +0x124
github.com/pingcap/tidb/pkg/executor.(*ExecStmt).Exec.func1()
        /Users/joechenrh/code/tidb-test/pkg/executor/adapter.go:484 +0x428
panic({0x1070259a0?, 0x10795bfe0?})
        /Users/joechenrh/deps/go/src/runtime/panic.go:770 +0x124
github.com/pingcap/tidb/pkg/infoschema.(*InfoschemaV3).TableByID(0x14006a72360, 0x6c)
        /Users/joechenrh/code/tidb-test/pkg/infoschema/infoschema_v3.go:225 +0xb8
github.com/pingcap/tidb/pkg/infoschema.(*SessionExtendedInfoSchema).TableByID(0x14007a63b90, 0x6c)
        /Users/joechenrh/code/tidb-test/pkg/infoschema/infoschema.go:843 +0x94
github.com/pingcap/tidb/pkg/executor.buildTableReq(0x14009931880, 0x3, {0x14005c8a870, 0x1, 0x14000da5cd8?})
        /Users/joechenrh/code/tidb-test/pkg/executor/builder.go:3913 +0x160
github.com/pingcap/tidb/pkg/executor.buildNoRangeIndexLookUpReader(0x14009931880, 0x14008425340)
        /Users/joechenrh/code/tidb-test/pkg/executor/builder.go:3981 +0xc4
github.com/pingcap/tidb/pkg/executor.(*executorBuilder).buildIndexLookUpReader(0x14009931880, 0x14008425340)
        /Users/joechenrh/code/tidb-test/pkg/executor/builder.go:4043 +0xa4
github.com/pingcap/tidb/pkg/executor.(*executorBuilder).build(0x10a1a80a0?, {0x1079cb4e0?, 0x14008425340})
        /Users/joechenrh/code/tidb-test/pkg/executor/builder.go:294 +0x98c
github.com/pingcap/tidb/pkg/executor.(*executorBuilder).buildProjection(0x14009931880, 0x1400a559b20)
        /Users/joechenrh/code/tidb-test/pkg/executor/builder.go:2013 +0x68
github.com/pingcap/tidb/pkg/executor.(*executorBuilder).build(0x107a23cd8?, {0x1079cafe0?, 0x1400a559b20})
        /Users/joechenrh/code/tidb-test/pkg/executor/builder.go:276 +0x92c
github.com/pingcap/tidb/pkg/executor.(*executorBuilder).buildExplain(0x14009931880, 0x140082ac8f0)
        /Users/joechenrh/code/tidb-test/pkg/executor/builder.go:1236 +0x1ac
github.com/pingcap/tidb/pkg/executor.(*executorBuilder).build(0x107a23cd8?, {0x1079cbe60?, 0x140082ac8f0})
        /Users/joechenrh/code/tidb-test/pkg/executor/builder.go:194 +0x119c
github.com/pingcap/tidb/pkg/executor.(*ExecStmt).buildExecutor(0x14006e12870)
        /Users/joechenrh/code/tidb-test/pkg/executor/adapter.go:1210 +0x164
github.com/pingcap/tidb/pkg/executor.(*ExecStmt).Exec(0x14006e12870, {0x10799fef8, 0x14008259ce0})
        /Users/joechenrh/code/tidb-test/pkg/executor/adapter.go:555 +0x688
github.com/pingcap/tidb/pkg/session.runStmt({0x10799fe50, 0x10ad5b848}, 0x14001994008, {0x1079b4640, 0x14006e12870})
        /Users/joechenrh/code/tidb-test/pkg/session/session.go:2288 +0x344
github.com/pingcap/tidb/pkg/session.(*session).ExecuteStmt(0x14001994008, {0x10799fe50?, 0x10ad5b848?}, {0x1079b9020, 0x14009931180})
        /Users/joechenrh/code/tidb-test/pkg/session/session.go:2149 +0xd64
github.com/pingcap/tidb/pkg/testkit.(*TestKit).ExecWithContext(0x140066a0050, {0x10799fe50, 0x10ad5b848}, {0x1400693aaa0, 0x44}, {0x0, 0x0, 0x0})
        /Users/joechenrh/code/tidb-test/pkg/testkit/testkit.go:385 +0x6f4
github.com/pingcap/tidb/pkg/testkit.(*TestKit).MustQueryWithContext(0x140066a0050, {0x10799fe50, 0x10ad5b848}, {0x1400693aaa0, 0x44}, {0x0, 0x0, 0x0})
        /Users/joechenrh/code/tidb-test/pkg/testkit/testkit.go:196 +0xc8
github.com/pingcap/tidb/pkg/testkit.(*TestKit).MustQuery(0x106341132?, {0x1400693aaa0?, 0x14000da7dc0?}, {0x0?, 0x1?, 0x0?})
        /Users/joechenrh/code/tidb-test/pkg/testkit/testkit.go:174 +0x78
github.com/pingcap/tidb/pkg/executor/test/tiflashtest.TestDisaggregatedTiFlashGeneratedColumn.func1(0x0)
        /Users/joechenrh/code/tidb-test/pkg/executor/test/tiflashtest/tiflash_test.go:1601 +0x2b4
github.com/pingcap/tidb/pkg/executor/test/tiflashtest.TestDisaggregatedTiFlashGeneratedColumn(0x1400193a9c0)
        /Users/joechenrh/code/tidb-test/pkg/executor/test/tiflashtest/tiflash_test.go:1668 +0x4f0
testing.tRunner(0x1400193a9c0, 0x10794df68)
        /Users/joechenrh/deps/go/src/testing/testing.go:1689 +0xec
created by testing.(*T).Run in goroutine 1
        /Users/joechenrh/deps/go/src/testing/testing.go:1742 +0x318
FAIL    github.com/pingcap/tidb/pkg/executor/test/tiflashtest   1.642s
```

### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->

--- a/pkg/ddl/index_cop.go
+++ b/pkg/ddl/index_cop.go
@@ -148,7 +148,7 @@ func buildDAGPB(exprCtx exprctx.BuildContext, distSQLCtx *distsqlctx.DistSQLCont
 }
 
 func constructTableScanPB(ctx exprctx.BuildContext, tblInfo *model.TableInfo, colInfos []*model.ColumnInfo) (*tipb.Executor, error) {
-	tblScan := tables.BuildTableScanFromInfos(tblInfo, colInfos)
+	tblScan := tables.BuildTableScanFromInfos(tblInfo, colInfos, false)
 	tblScan.TableId = tblInfo.ID
 	err := tables.SetPBColumnsDefaultValue(ctx, tblScan.Columns, colInfos)
 	return &tipb.Executor{Tp: tipb.ExecType_TypeTableScan, TblScan: tblScan}, err

--- a/pkg/ddl/reorg.go
+++ b/pkg/ddl/reorg.go
@@ -521,7 +521,7 @@ func (r *reorgInfo) String() string {
 }
 
 func constructDescTableScanPB(physicalTableID int64, tblInfo *model.TableInfo, handleCols []*model.ColumnInfo) *tipb.Executor {
-	tblScan := tables.BuildTableScanFromInfos(tblInfo, handleCols)
+	tblScan := tables.BuildTableScanFromInfos(tblInfo, handleCols, false)
 	tblScan.TableId = physicalTableID
 	tblScan.Desc = true
 	return &tipb.Executor{Tp: tipb.ExecType_TypeTableScan, TblScan: tblScan}

--- a/pkg/executor/admin.go
+++ b/pkg/executor/admin.go
@@ -163,7 +163,7 @@ func (e *CheckIndexRangeExec) constructIndexScanPB() *tipb.Executor {
 	idxExec := &tipb.IndexScan{
 		TableId: e.table.ID,
 		IndexId: e.index.ID,
-		Columns: util.ColumnsToProto(e.cols, e.table.PKIsHandle, true),
+		Columns: util.ColumnsToProto(e.cols, e.table.PKIsHandle, true, false),
 	}
 	return &tipb.Executor{Tp: tipb.ExecType_TypeIndexScan, IdxScan: idxExec}
 }
@@ -228,7 +228,7 @@ func (e *RecoverIndexExec) Open(ctx context.Context) error {
 }
 
 func (e *RecoverIndexExec) constructTableScanPB(tblInfo *model.TableInfo, colInfos []*model.ColumnInfo) (*tipb.Executor, error) {
-	tblScan := tables.BuildTableScanFromInfos(tblInfo, colInfos)
+	tblScan := tables.BuildTableScanFromInfos(tblInfo, colInfos, false)
 	tblScan.TableId = e.physicalID
 	err := tables.SetPBColumnsDefaultValue(e.Ctx().GetExprCtx(), tblScan.Columns, colInfos)
 	return &tipb.Executor{Tp: tipb.ExecType_TypeTableScan, TblScan: tblScan}, err
@@ -880,7 +880,7 @@ func (e *CleanupIndexExec) constructIndexScanPB() *tipb.Executor {
 	idxExec := &tipb.IndexScan{
 		TableId:          e.physicalID,
 		IndexId:          e.index.Meta().ID,
-		Columns:          util.ColumnsToProto(e.columns, e.table.Meta().PKIsHandle, true),
+		Columns:          util.ColumnsToProto(e.columns, e.table.Meta().PKIsHandle, true, false),
 		PrimaryColumnIds: tables.TryGetCommonPkColumnIds(e.table.Meta()),
 	}
 	return &tipb.Executor{Tp: tipb.ExecType_TypeIndexScan, IdxScan: idxExec}

--- a/pkg/executor/builder.go
+++ b/pkg/executor/builder.go
@@ -2881,7 +2881,7 @@ func (b *executorBuilder) buildAnalyzeSamplingPushdown(
 		SampleSize:   int64(opts[ast.AnalyzeOptNumSamples]),
 		SampleRate:   sampleRate,
 		SketchSize:   statistics.MaxSketchSize,
-		ColumnsInfo:  util.ColumnsToProto(task.ColsInfo, task.TblInfo.PKIsHandle, false),
+		ColumnsInfo:  util.ColumnsToProto(task.ColsInfo, task.TblInfo.PKIsHandle, false, false),
 		ColumnGroups: colGroups,
 	}
 	if task.TblInfo != nil {
@@ -3011,7 +3011,7 @@ func (b *executorBuilder) buildAnalyzeColumnsPushdown(
 		BucketSize:    int64(opts[ast.AnalyzeOptNumBuckets]),
 		SampleSize:    MaxRegionSampleSize,
 		SketchSize:    statistics.MaxSketchSize,
-		ColumnsInfo:   util.ColumnsToProto(cols, task.HandleCols != nil && task.HandleCols.IsInt(), false),
+		ColumnsInfo:   util.ColumnsToProto(cols, task.HandleCols != nil && task.HandleCols.IsInt(), false, false),
 		CmsketchDepth: &depth,
 		CmsketchWidth: &width,
 	}

--- a/pkg/planner/core/find_best_task.go
+++ b/pkg/planner/core/find_best_task.go
@@ -2458,13 +2458,7 @@ func (ds *DataSource) convertToTableScan(prop *property.PhysicalProperty, candid
 		// TiFlash fast mode(https://github.com/pingcap/tidb/pull/35851) does not keep order in TableScan
 		return base.InvalidTask, nil
 	}
-	if ts.StoreType == kv.TiFlash {
-		for _, col := range ts.Columns {
-			if col.IsVirtualGenerated() {
-				col.AddFlag(mysql.GeneratedColumnFlag)
-			}
-		}
-	}
+
 	// In disaggregated tiflash mode, only MPP is allowed, cop and batchCop is deprecated.
 	// So if prop.TaskTp is RootTaskType, have to use mppTask then convert to rootTask.
 	isTiFlashPath := ts.StoreType == kv.TiFlash
@@ -2814,7 +2808,7 @@ func (ts *PhysicalTableScan) getScanRowSize() float64 {
 func (ds *DataSource) getOriginalPhysicalTableScan(prop *property.PhysicalProperty, path *util.AccessPath, isMatchProp bool) (*PhysicalTableScan, float64) {
 	ts := PhysicalTableScan{
 		Table:           ds.TableInfo,
-		Columns:         util.CloneColInfos(ds.Columns),
+		Columns:         slices.Clone(ds.Columns),
 		TableAsName:     ds.TableAsName,
 		DBName:          ds.DBName,
 		isPartition:     ds.PartitionDefIdx != nil,

--- a/pkg/planner/core/find_best_task.go
+++ b/pkg/planner/core/find_best_task.go
@@ -2814,7 +2814,7 @@ func (ts *PhysicalTableScan) getScanRowSize() float64 {
 func (ds *DataSource) getOriginalPhysicalTableScan(prop *property.PhysicalProperty, path *util.AccessPath, isMatchProp bool) (*PhysicalTableScan, float64) {
 	ts := PhysicalTableScan{
 		Table:           ds.TableInfo,
-		Columns:         slices.Clone(ds.Columns),
+		Columns:         util.CloneColInfos(ds.Columns),
 		TableAsName:     ds.TableAsName,
 		DBName:          ds.DBName,
 		isPartition:     ds.PartitionDefIdx != nil,

--- a/pkg/planner/core/plan_to_pb.go
+++ b/pkg/planner/core/plan_to_pb.go
@@ -258,7 +258,7 @@ func (p *PhysicalTableScan) ToPB(ctx *base.BuildPBContext, storeType kv.StoreTyp
 	if storeType == kv.TiFlash && p.Table.GetPartitionInfo() != nil && p.IsMPPOrBatchCop && p.SCtx().GetSessionVars().StmtCtx.UseDynamicPartitionPrune() {
 		return p.partitionTableScanToPBForFlash(ctx)
 	}
-	tsExec := tables.BuildTableScanFromInfos(p.Table, p.Columns)
+	tsExec := tables.BuildTableScanFromInfos(p.Table, p.Columns, p.StoreType == kv.TiFlash)
 	tsExec.Desc = p.Desc
 	keepOrder := p.KeepOrder
 	tsExec.KeepOrder = &keepOrder
@@ -488,7 +488,7 @@ func (p *PhysicalIndexScan) ToPB(_ *base.BuildPBContext, _ kv.StoreType) (*tipb.
 	idxExec := &tipb.IndexScan{
 		TableId:          p.Table.ID,
 		IndexId:          p.Index.ID,
-		Columns:          util.ColumnsToProto(columns, p.Table.PKIsHandle, true),
+		Columns:          util.ColumnsToProto(columns, p.Table.PKIsHandle, true, false),
 		Desc:             p.Desc,
 		PrimaryColumnIds: pkColIDs,
 	}

--- a/pkg/planner/core/plan_to_pb.go
+++ b/pkg/planner/core/plan_to_pb.go
@@ -22,7 +22,6 @@ import (
 	"github.com/pingcap/tidb/pkg/expression/aggregation"
 	"github.com/pingcap/tidb/pkg/kv"
 	"github.com/pingcap/tidb/pkg/parser/model"
-	"github.com/pingcap/tidb/pkg/parser/mysql"
 	"github.com/pingcap/tidb/pkg/planner/core/base"
 	"github.com/pingcap/tidb/pkg/planner/core/operator/logicalop"
 	util2 "github.com/pingcap/tidb/pkg/planner/util"
@@ -272,16 +271,6 @@ func (p *PhysicalTableScan) ToPB(ctx *base.BuildPBContext, storeType kv.StoreTyp
 			return nil, err
 		}
 		tsExec.PushedDownFilterConditions = conditions
-	}
-
-	if p.StoreType == kv.TiFlash {
-		for _, col := range p.Columns {
-			if col.IsVirtualGenerated() {
-				// The column may be shallow copied from info cache, so make a copy before modify it.
-				col = col.Clone()
-				col.AddFlag(mysql.GeneratedColumnFlag)
-			}
-		}
 	}
 
 	var err error

--- a/pkg/planner/core/plan_to_pb.go
+++ b/pkg/planner/core/plan_to_pb.go
@@ -22,6 +22,7 @@ import (
 	"github.com/pingcap/tidb/pkg/expression/aggregation"
 	"github.com/pingcap/tidb/pkg/kv"
 	"github.com/pingcap/tidb/pkg/parser/model"
+	"github.com/pingcap/tidb/pkg/parser/mysql"
 	"github.com/pingcap/tidb/pkg/planner/core/base"
 	"github.com/pingcap/tidb/pkg/planner/core/operator/logicalop"
 	util2 "github.com/pingcap/tidb/pkg/planner/util"
@@ -271,6 +272,16 @@ func (p *PhysicalTableScan) ToPB(ctx *base.BuildPBContext, storeType kv.StoreTyp
 			return nil, err
 		}
 		tsExec.PushedDownFilterConditions = conditions
+	}
+
+	if p.StoreType == kv.TiFlash {
+		for _, col := range p.Columns {
+			if col.IsVirtualGenerated() {
+				// The column may be shallow copied from info cache, so make a copy before modify it.
+				col = col.Clone()
+				col.AddFlag(mysql.GeneratedColumnFlag)
+			}
+		}
 	}
 
 	var err error

--- a/pkg/planner/core/plan_to_pb_test.go
+++ b/pkg/planner/core/plan_to_pb_test.go
@@ -35,16 +35,16 @@ func TestColumnToProto(t *testing.T) {
 	col := &model.ColumnInfo{
 		FieldType: *tp,
 	}
-	pc := util.ColumnToProto(col, false)
+	pc := util.ColumnToProto(col, false, false)
 	expect := &tipb.ColumnInfo{ColumnId: 0, Tp: 3, Collation: 83, ColumnLen: 11, Decimal: 0, Flag: 10, Elems: []string(nil), DefaultVal: []uint8(nil), PkHandle: false, XXX_unrecognized: []uint8(nil)}
 	require.Equal(t, expect, pc)
 
 	cols := []*model.ColumnInfo{col, col}
-	pcs := util.ColumnsToProto(cols, false, false)
+	pcs := util.ColumnsToProto(cols, false, false, false)
 	for _, v := range pcs {
 		require.Equal(t, int32(10), v.GetFlag())
 	}
-	pcs = util.ColumnsToProto(cols, true, false)
+	pcs = util.ColumnsToProto(cols, true, false, false)
 	for _, v := range pcs {
 		require.Equal(t, int32(10), v.GetFlag())
 	}
@@ -56,19 +56,19 @@ func TestColumnToProto(t *testing.T) {
 	col1 := &model.ColumnInfo{
 		FieldType: *tp,
 	}
-	pc = util.ColumnToProto(col1, false)
+	pc = util.ColumnToProto(col1, false, false)
 	require.Equal(t, int32(8), pc.Collation)
 
 	collate.SetNewCollationEnabledForTest(true)
 
-	pc = util.ColumnToProto(col, false)
+	pc = util.ColumnToProto(col, false, false)
 	expect = &tipb.ColumnInfo{ColumnId: 0, Tp: 3, Collation: -83, ColumnLen: 11, Decimal: 0, Flag: 10, Elems: []string(nil), DefaultVal: []uint8(nil), PkHandle: false, XXX_unrecognized: []uint8(nil)}
 	require.Equal(t, expect, pc)
-	pcs = util.ColumnsToProto(cols, true, false)
+	pcs = util.ColumnsToProto(cols, true, false, false)
 	for _, v := range pcs {
 		require.Equal(t, int32(-83), v.Collation)
 	}
-	pc = util.ColumnToProto(col1, false)
+	pc = util.ColumnToProto(col1, false, false)
 	require.Equal(t, int32(-8), pc.Collation)
 
 	tp = types.NewFieldType(mysql.TypeEnum)
@@ -77,7 +77,7 @@ func TestColumnToProto(t *testing.T) {
 	col2 := &model.ColumnInfo{
 		FieldType: *tp,
 	}
-	pc = util.ColumnToProto(col2, false)
+	pc = util.ColumnToProto(col2, false, false)
 	require.Len(t, pc.Elems, 2)
 
 	tp = types.NewFieldTypeBuilder().
@@ -91,7 +91,7 @@ func TestColumnToProto(t *testing.T) {
 	col3 := &model.ColumnInfo{
 		FieldType: *tp,
 	}
-	pc = util.ColumnToProto(col3, true)
+	pc = util.ColumnToProto(col3, true, false)
 	expect = &tipb.ColumnInfo{ColumnId: 0, Tp: 0xfe, Collation: 63, ColumnLen: 100, Decimal: 0, Flag: 10, Elems: []string(nil), DefaultVal: []uint8(nil), PkHandle: false, XXX_unrecognized: []uint8(nil)}
 	require.Equal(t, expect, pc)
 }

--- a/pkg/table/tables/tables.go
+++ b/pkg/table/tables/tables.go
@@ -1953,11 +1953,11 @@ func getSequenceAllocator(allocs autoid.Allocators) (autoid.Allocator, error) {
 }
 
 // BuildTableScanFromInfos build tipb.TableScan with *model.TableInfo and *model.ColumnInfo.
-func BuildTableScanFromInfos(tableInfo *model.TableInfo, columnInfos []*model.ColumnInfo) *tipb.TableScan {
+func BuildTableScanFromInfos(tableInfo *model.TableInfo, columnInfos []*model.ColumnInfo, isTiFlashStore bool) *tipb.TableScan {
 	pkColIDs := TryGetCommonPkColumnIds(tableInfo)
 	tsExec := &tipb.TableScan{
 		TableId:          tableInfo.ID,
-		Columns:          util.ColumnsToProto(columnInfos, tableInfo.PKIsHandle, false),
+		Columns:          util.ColumnsToProto(columnInfos, tableInfo.PKIsHandle, false, isTiFlashStore),
 		PrimaryColumnIds: pkColIDs,
 	}
 	if tableInfo.IsCommonHandle {
@@ -1971,7 +1971,7 @@ func BuildPartitionTableScanFromInfos(tableInfo *model.TableInfo, columnInfos []
 	pkColIDs := TryGetCommonPkColumnIds(tableInfo)
 	tsExec := &tipb.PartitionTableScan{
 		TableId:          tableInfo.ID,
-		Columns:          util.ColumnsToProto(columnInfos, tableInfo.PKIsHandle, false),
+		Columns:          util.ColumnsToProto(columnInfos, tableInfo.PKIsHandle, false, false),
 		PrimaryColumnIds: pkColIDs,
 		IsFastScan:       &fastScan,
 	}

--- a/pkg/util/misc.go
+++ b/pkg/util/misc.go
@@ -394,10 +394,10 @@ func TLSCipher2String(n uint16) string {
 }
 
 // ColumnsToProto converts a slice of model.ColumnInfo to a slice of tipb.ColumnInfo.
-func ColumnsToProto(columns []*model.ColumnInfo, pkIsHandle bool, forIndex bool) []*tipb.ColumnInfo {
+func ColumnsToProto(columns []*model.ColumnInfo, pkIsHandle bool, forIndex bool, isTiFlashStore bool) []*tipb.ColumnInfo {
 	cols := make([]*tipb.ColumnInfo, 0, len(columns))
 	for _, c := range columns {
-		col := ColumnToProto(c, forIndex)
+		col := ColumnToProto(c, forIndex, isTiFlashStore)
 		// TODO: Here `PkHandle`'s meaning is changed, we will change it to `IsHandle` when tikv's old select logic
 		// is abandoned.
 		if (pkIsHandle && mysql.HasPriKeyFlag(c.GetFlag())) || c.ID == model.ExtraHandleID {
@@ -411,7 +411,7 @@ func ColumnsToProto(columns []*model.ColumnInfo, pkIsHandle bool, forIndex bool)
 }
 
 // ColumnToProto converts model.ColumnInfo to tipb.ColumnInfo.
-func ColumnToProto(c *model.ColumnInfo, forIndex bool) *tipb.ColumnInfo {
+func ColumnToProto(c *model.ColumnInfo, forIndex bool, isTiFlashStore bool) *tipb.ColumnInfo {
 	pc := &tipb.ColumnInfo{
 		ColumnId:  c.ID,
 		Collation: collate.RewriteNewCollationIDIfNeeded(int32(mysql.CollationNames[c.GetCollate()])),
@@ -420,7 +420,7 @@ func ColumnToProto(c *model.ColumnInfo, forIndex bool) *tipb.ColumnInfo {
 		Flag:      int32(c.GetFlag()),
 		Elems:     c.GetElems(),
 	}
-	if c.IsVirtualGenerated() {
+	if isTiFlashStore && c.IsVirtualGenerated() {
 		pc.Flag |= int32(mysql.GeneratedColumnFlag)
 	}
 	if forIndex {

--- a/pkg/util/misc.go
+++ b/pkg/util/misc.go
@@ -420,6 +420,9 @@ func ColumnToProto(c *model.ColumnInfo, forIndex bool) *tipb.ColumnInfo {
 		Flag:      int32(c.GetFlag()),
 		Elems:     c.GetElems(),
 	}
+	if c.IsVirtualGenerated() {
+		pc.Flag |= int32(mysql.GeneratedColumnFlag)
+	}
 	if forIndex {
 		// Use array type for read the multi-valued index.
 		pc.Tp = int32(c.FieldType.ArrayType().GetType())

--- a/pkg/util/misc_test.go
+++ b/pkg/util/misc_test.go
@@ -174,8 +174,8 @@ func TestToPB(t *testing.T) {
 	}
 	column2.SetCollate("utf8mb4_bin")
 
-	assert.Equal(t, "column_id:1 collation:-45 columnLen:-1 decimal:-1 ", ColumnToProto(column, false).String())
-	assert.Equal(t, "column_id:1 collation:-45 columnLen:-1 decimal:-1 ", ColumnsToProto([]*model.ColumnInfo{column, column2}, false, false)[0].String())
+	assert.Equal(t, "column_id:1 collation:-45 columnLen:-1 decimal:-1 ", ColumnToProto(column, false, false).String())
+	assert.Equal(t, "column_id:1 collation:-45 columnLen:-1 decimal:-1 ", ColumnsToProto([]*model.ColumnInfo{column, column2}, false, false, false)[0].String())
 }
 
 func TestComposeURL(t *testing.T) {

